extern/storage-sealing/input.go

```diff
@@ -0,0 +1,424 @@
+package sealing
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/ipfs/go-cid"
+
+	"github.com/filecoin-project/go-padreader"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-statemachine"
+	"github.com/filecoin-project/specs-storage/storage"
+
+	sectorstorage "github.com/filecoin-project/lotus/extern/sector-storage"
+	"github.com/filecoin-project/lotus/extern/sector-storage/ffiwrapper"
+)
+
+func (m *Sealing) handleWaitDeals(ctx statemachine.Context, sector SectorInfo) error {
+	var used abi.UnpaddedPieceSize
+	for _, piece := range sector.Pieces {
+		used += piece.Piece.Size.Unpadded()
+	}
+
+	m.inputLk.Lock()
+
+	started, err := m.maybeStartSealing(ctx, sector, used)
+	if err != nil || started {
+		delete(m.openSectors, m.minerSectorID(sector.SectorNumber))
+
+		m.inputLk.Unlock()
+
+		return err
+	}
+
+	m.openSectors[m.minerSectorID(sector.SectorNumber)] = &openSector{
+		used: used,
+		maybeAccept: func(cid cid.Cid) error {
+			// todo check deal start deadline (configurable)
+
+			sid := m.minerSectorID(sector.SectorNumber)
+			m.assignedPieces[sid] = append(m.assignedPieces[sid], cid)
+
+			return ctx.Send(SectorAddPiece{})
+		},
+	}
+
+	go func() {
+		defer m.inputLk.Unlock()
+		if err := m.updateInput(ctx.Context(), sector.SectorType); err != nil {
+			log.Errorf("%+v", err)
+		}
+	}()
+
+	return nil
+}
+
+func (m *Sealing) maybeStartSealing(ctx statemachine.Context, sector SectorInfo, used abi.UnpaddedPieceSize) (bool, error) {
+	now := time.Now()
+	st := m.sectorTimers[m.minerSectorID(sector.SectorNumber)]
+	if st != nil {
+		if !st.Stop() { // timer expired, SectorStartPacking was/is being sent
+			// we send another SectorStartPacking in case one was sent in the handleAddPiece state
+			log.Infow("starting to seal deal sector", "sector", sector.SectorNumber, "trigger", "wait-timeout")
+			return true, ctx.Send(SectorStartPacking{})
+		}
+	}
+
+	ssize, err := sector.SectorType.SectorSize()
+	if err != nil {
+		return false, xerrors.Errorf("getting sector size")
+	}
+
+	maxDeals, err := getDealPerSectorLimit(ssize)
+	if err != nil {
+		return false, xerrors.Errorf("getting per-sector deal limit: %w", err)
+	}
+
+	if len(sector.dealIDs()) >= maxDeals {
+		// can't accept more deals
+		log.Infow("starting to seal deal sector", "sector", sector.SectorNumber, "trigger", "maxdeals")
+		return true, ctx.Send(SectorStartPacking{})
+	}
+
+	if used.Padded() == abi.PaddedPieceSize(ssize) {
```

Comment:
This should probably have some leeway, there is no reason to wait for the timeout because there is room for a 10byte deal.

Reply:
Well, you,ll never have less than the minimum deal size free, which is already configurable - but yes, we could add a separate config for this